### ISSUE TITLE
minimumSize=1 generated for choice that contains child with zero size

### ISSUE THREAD

**Issue by zhztheplayer** [javacc/javacc#86]
Following code lines seem to cause lookahead problems:

https://github.com/javacc/javacc/blob/ca275162a5d98d3bd4f2317efd9c19ad2144b5ef/src/main/java/org/javacc/parser/ParseEngine.java#L1377-L1381

The minimum consumed size will be limited to at least 1 for a choice. As a result sometimes the lookahead is finished too early. E.g. consider an production:

```
Syntax() -> {
  (
  LOOKAHEAD(2) 
  (<FOO> | {}) <BAR> <DUMMY>
  |
  <BAR> <BAR>
  )
}
```

If the input tokens is `<BAR> <BAR>`, apparently we should choose `<BAR> <BAR>` because of the lookahead for the former expansion is specified to 2. But with the code logic pasted above, the expansion `[<FOO>] <BAR>` will consume all the 2 lookahead count so that the generated parser code will never consider the token `<DUMMY>`. After all the former expansion will be chosen incorrectly. :(

To me it is definitely like a bug so I have opened a PR #87. If it is something by design (which means we should always use lookahead=3 in the case) please let me know.

**Comment by julianhyde** [comment]
This issue is affecting Apache Calcite's SQL parser (specifically, our efforts to reduce default lookahead from 2 to 1). See https://issues.apache.org/jira/browse/CALCITE-2847.

We have a workaround but would be grateful if the fix could be included in the next version of JavaCC.

**Comment by new-javacc** [comment]
OK. I think you guys misunderstood what LOOKAHEAD(2) means. It means max lookahead 2. Since you have Empty in the first choice, that will trivially match and then there is BAR so it's correct. This bug is invaid and please undo the change.

**Comment by new-javacc** [comment]
This bug is invalid.

**Comment by new-javacc** [comment]
Also note that JavaCC takes choices in order they are in the input grammar.

**Comment by zhztheplayer** [comment]
Thanks for your reply, @new-javacc!

I am totally with you with the fact that "LOOKAHEAD(2) means max lookahead 2", but unfortunately I am still confident with the existing of the issue. :(

LOOKAHEAD(x) means JavaCC should scan on no more than x tokens, I think this is definitely by the design of JavaCC fundamental. But the issue is not about the runtime execution of lookahead, it's about whether we generated enough `jj_scan_token(...)` codes.

Let's see the issue by another facet. JavaCC generates the zero minimum size 0 for the "zero or one" expansion as following:

https://github.com/javacc/javacc/blob/ca275162a5d98d3bd4f2317efd9c19ad2144b5ef/src/main/java/org/javacc/parser/ParseEngine.java#L1408-L1409

I we can come in to agreement that choice expansion `(<FOO>|{})` and "zero or one" expansion `[<FOO>]` are literally equivalent, that is saying, the JavaCC would behave differently on such expansions. That is saying if I use the code:

```
Syntax() -> {
  (
  LOOKAHEAD(2) 
  [<FOO>] <BAR> <DUMMY>
  |
  <BAR> <BAR>
  )
}
```
I will never meet the issue. AFAICS the behavior is apparently inconsistent with the behavior of the code pasted in the first comment.

Besides, I use the simplest case for demonstrating the issue. In production, the use of such a choice expansion might be hidden in very deep nested productions so that we'll hardly choose a correct lookahead hint number without debugging the parser code.

TL;DR

LOOKAHEAD(2) means max lookahead 2, but JavaCC does not generate enough code for consuming 2 lookahead counts.

**Comment by new-javacc** [comment]
OK but one more thing you should remember is that {} is NOT empty. 
Action block is an expansion unit with no lookahead. So (<FOO> | {} ) is 
not same as [ <FOO> ], I don't think. But let me double check tomorrow.

**Comment by zhztheplayer** [comment]
Thank you very much @new-javacc ! Looking forward to any feedback on this.

**Comment by zhztheplayer** [comment]
@new-javacc - Is there any progress about this? It would be good if we can reopen this issue so that we don't forget resolving it.

**Comment by new-javacc** [comment]
I thought I mentioned above - it's not a bug. An empty action is alway taken because the lookahead is trivially true. Again like I said:

[ x ]

is not the same as

{} | x

**Comment by zhztheplayer** [comment]
@new-javacc , the problem is not really related to actions. It is the problem of the way we generate lookahead codes for **Choice** is too optimistic. I might use `( <FOO> | {} )` as an example, but it is only because the empty action consumes 0 lookahead count in minimal[1]. The same issue happens on `( <FOO> | [<DUMMY>] )`, `( <FOO> | (<DUMMY>)* )`, whatever as long as one of the choices is treated **0**.

[1] [https://<span></span>github.com/javacc/javacc/blob/ca275162a5d98d3bd4f2317efd9c19ad2144b5ef/src/main/java/org/javacc/parser/ParseEngine.java#L1413](https://github.com/javacc/javacc/blob/ca275162a5d98d3bd4f2317efd9c19ad2144b5ef/src/main/java/org/javacc/parser/ParseEngine.java#L1413)

**Comment by zhztheplayer** [comment]
Because of the broken loop condition `min > 1` in the code, if the first branch of a choice is minimally 2, then nothing goes wrong, say we have syntax: 
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> <FOO> | {} ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```

The generated code is good for parsing <BAR> <BAR>. If we make the branch consumes 1 minimally (it is the case I used in PR):

```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> | {} ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```

The error will be thrown this time. It's all because a simple mistake (I think) when someone writes the following loop: 
https://github.com/javacc/javacc/blob/ca275162a5d98d3bd4f2317efd9c19ad2144b5ef/src/main/java/org/javacc/parser/ParseEngine.java#L1377-L1381

The condition `min > 1` causes loop breaks iff it sees a **1** or **0**,  but not **2**. It doesn't makes sense because the one nexts to **2** might happen to be an **0** (see the first example in this comment). The logic is counter-intuitive and I am pretty sure it is not something by design.

**Comment by new-javacc** [comment]
Like I said before - I want the grammar and the input. I don't really 
want to talk about the generated code. What is  the input string that 
shows the "wrong parse" for the cases. I think you are misunderstanding 
how JavaCC lookahead works. Did you look at the tutorial(s)? There are 
some good examples there.

**Comment by new-javacc** [comment]
Better yet - generate the parser with DEBUG_LOKAHEAD option and see what 
it's doing.

**Comment by zhztheplayer** [comment]
Thanks @new-javacc. I was talking about the generated code because at least for the method ParseEngine#minimumSize, I know what it intends to do and how it works. I read the [lookahead document](https://javacc.org/tutorials/lookahead) for times so per my experience this problem should not be intended or by design.

And I think I already provide some other cases in comment https://github.com/javacc/javacc/issues/86#issuecomment-495965072. I can list them below in detail (will add to PR if needed):

---

### Bad Case 1 (in the PR)
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> | {} ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
ParseException: Encountered " "BAR" "BAR" " at line 1, column 5.

---

### Good Case 1
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> <FOO> | {} ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
(Successfully parsed, no output)

---

### Bad Case 2
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> | [<DUMMY>] ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
ParseException: Encountered " "BAR" "BAR" " at line 1, column 5.

---

### Good Case 2
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> <FOO> | [<DUMMY>] ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
(Successfully parsed, no output)

---

### Bad Case 3
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> | (<DUMMY>)* ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
ParseException: Encountered " "BAR" "BAR" " at line 1, column 5.

---

### Good Case 3
#### Syntax
```
void CompilationUnit() : {}
{
	LOOKAHEAD(2)
	( <FOO> <FOO> | (<DUMMY>)* ) <BAR> <DUMMY>
	|
	<BAR> <BAR>
}
```
#### Input
BAR BAR
#### Expected output
(Successfully parsed, no output)
#### Actual output
(Successfully parsed, no output)

---

@new-javacc, could you please take a look again ? Thank you very much.

**Comment by zhztheplayer** [comment]
> An empty action is alway taken because the lookahead is trivially true.

@new-javacc , also, the lookahead for empty action is true, but the action does't consume lookahead counts. Say:
* Grammar `({} <FOO> | <BAR> )` is good for parsing input `BAR`;
* Grammar `(( <DUMMY> | {} ) <FOO> | <BAR> )` is good for parsing input `BAR`; 
* HOWEVER grammer `( LOOKAHEAD(2) ( <DUMMY> | {} ) <BAR> <DUMMY> | <BAR> <BAR> )` is not able to parse input `BAR BAR`. 

This is simply INCORRECT.

**Comment by new-javacc** [comment]
Like I said, it's not incorrect. There are some really subtle rules. Let 
me read carefully the lookahead mini tutorial and get back to you. 
Remember the 2 rules - 1. LOOKEHAD is the max we want it to use so it's 
free to use anything upto that and 2.  javacc does not do lookahead 
during lookahead. As for your examples, the number of choice points are 
different so you can't  expect the same behaviour. Only thing I would 
say if a rule works without lookahead, then should also work with 
lookahead.

So I'm still not convinced there is a bug. You can always reorder your 
rules to put <BAR><BAR> first - that's standard practice in javacc. It 
simply takes the first choice at a choice point that's feasible due to 
implicit/explicit lookahead. It does not try all possibilites.

In fact, I say keep the bug open as low priority and use the workaround 
of reordering the rules.

**Comment by new-javacc** [comment]
Adding javacc users  so more people can chime in.

**Comment by new-javacc** [comment]
Thanks for the details. There are a lot of subtle semantic issues in 
JavaCC lookahead and I generally don't expect any bug there. So in all 
the good cases if you see - there is *at least* one choice that needs a 
lookahead of 2 so the lookahead(2) makes sense. In all your bad cases, 
all the choices are of size <= 1 so lookahead 2 has no effect. So the 
"bad cases" you mentioned below are not actually bugs. Run it with 
DEBUG_LOOKAHEAD and it will tell you exactly what it is doing. Just to 
reiterate LOOKAHEAD(2) says upto two tokens but if the choices are less 
than 2 tokens long, it will not use 2.

Hope that helps.

**Comment by zhztheplayer** [comment]
Thanks @new-javacc .
> Just to reiterate LOOKAHEAD(2) says upto two tokens but if the choices are less than 2 tokens long, it will not use 2.

Sorry I don't think so (If I understand your point correctly). For example, I can write following grammars:

1. `( LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )`
2. `( ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )`

By your explanation, the 2 grammars should be effectively equivalent, because the leading expansion ( \<FOO> | \<BAR> ) is a choice with no longer than 2 tokens. However, grammar 1 can parse `BAR BAR` correctly, where grammar 2 cannot. I think the behavior is right.

Per lookahead tutorial and JavaCC generated code, I believe a lookahead hint in choice should be the lookahead for the whole choice branch, not only the first expansion. Say, the lookahead in `LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | ...` should consider `( <FOO> | <BAR> ) <DUMMY>`, not only the sub choice `( <FOO> | <BAR> )`, so I don't think length of the sub choice matters.

**Comment by new-javacc** [comment]
Yeas that's because the default lookahead is 1On May 31, 2019 2:51 AM, Hongze Zhang <notifications@github.com> wrote:Thanks @new-javacc .

Just to reiterate LOOKAHEAD(2) says upto two tokens but if the choices are less than 2 tokens long, it will not use 2.

Sorry I don't think so (If I understand your point correctly). For example, I can write following grammars:
( LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )( ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )
By your explanation, the 2 grammars should be effectively equivalent, because the leading expansion ( <FOO> | <BAR> ) is a choice with no longer than 2 tokens. However, grammar 1 can parse BAR BAR correctly, where grammar 2 cannot. I think the behavior is right.
Per lookahead tutorial and JavaCC generated code, I believe a lookahead hint in choice should be the lookahead for the whole choice branch, not only the first expansion. Say, the lookahead in LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | ... should consider ( <FOO> | <BAR> ) <DUMMY>, not only the sub choice ( <FOO> | <BAR> ), so I don't think length of the sub choice matters.

**Comment by new-javacc** [comment]
Oops hit send too soon.Basically your example perfectly illustrates the need for lookahead. So in your second example below sicne the default lookahead is 1, it will see that bar matches in the first choice and so it commits to that. So the two examples are *not* equivalent.I strongly suggest you run the grammar with debug_lookahead option and carefully see the output to understand what it's doing.On May 31, 2019 9:23 AM, JavaCC Support <support@javacc.org> wrote:Yeas that's because the default lookahead is 1On May 31, 2019 2:51 AM, Hongze Zhang <notifications@github.com> wrote:Thanks @new-javacc .

Just to reiterate LOOKAHEAD(2) says upto two tokens but if the choices are less than 2 tokens long, it will not use 2.

Sorry I don't think so (If I understand your point correctly). For example, I can write following grammars:
( LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )( ( <FOO> | <BAR> ) <DUMMY> | <BAR> <BAR> )
By your explanation, the 2 grammars should be effectively equivalent, because the leading expansion ( <FOO> | <BAR> ) is a choice with no longer than 2 tokens. However, grammar 1 can parse BAR BAR correctly, where grammar 2 cannot. I think the behavior is right.
Per lookahead tutorial and JavaCC generated code, I believe a lookahead hint in choice should be the lookahead for the whole choice branch, not only the first expansion. Say, the lookahead in LOOKAHEAD(2) ( <FOO> | <BAR> ) <DUMMY> | ... should consider ( <FOO> | <BAR> ) <DUMMY>, not only the sub choice ( <FOO> | <BAR> ), so I don't think length of the sub choice matters.

**Comment by zhztheplayer** [comment]
@new-javacc , I think I used to debug the syntax by turning DEBUG_LOOKAHEAD on several month ago, I think I already now the parser-gen process for such grammars but anyway I'll run again and paste the output here soon.

> So in your second example below sicne the default lookahead is 1, it will see that bar matches in the first choice and so it commits to that. So the two examples are *not* equivalent.

I totally agree with you. But in your earlier comment you said "LOOKAHEAD(2) says upto two tokens but if the choices are less than 2 tokens long, it will not use 2". Say we have 4 grammars: 

1. ( LOOKAHEAD(2) ( \<FOO> | {} ) \<BAR> \<DUMMY> | \<BAR> \<BAR>)
2. ( ( \<FOO> | {} ) \<BAR> \<DUMMY> | \<BAR> \<BAR> )
3. ( LOOKAHEAD(2) ( \<FOO> | \<BAR> ) \<DUMMY> | \<BAR> \<BAR> )
4. ( ( \<FOO> | \<BAR> ) \<DUMMY> | \<BAR> \<BAR> )

Do you mean that 3 and 4 are not equivalent but 1 and 2 are equivalent? Grammar 3 can parse `BAR BAR`, where grammar 1 cannot parse `BAR BAR`. Both subchoices are of <=1 long. Could you please explain a bit more about the reason except for there is a bug?

**Comment by zhztheplayer** [comment]
The debug output for following grammar and input:
- grammar
```
void CompilationUnit() : {}
{
    ( LOOKAHEAD(2) Branch1() | Branch2() )
}

void Branch1() : {}
{
    ( <FOO> | {} ) <BAR> <DUMMY>
}

void Branch2() : {}
{
    <BAR> <BAR>
}
```
- input
BAR BAR

- debug output
```
Call:	CompilationUnit
  Call:	Branch1(LOOKING AHEAD...)
    Visited token: <"BAR" at line 1 column 1>; Expected token: <"FOO">
    Visited token: <"BAR" at line 1 column 1>; Expected token: <"BAR">
  Return: Branch1(LOOKAHEAD SUCCEEDED)
  Call:	Branch1
    Consumed token: <"BAR" at line 1 column 1>
  Return: Branch1
Return: CompilationUnit
Exception in thread "main" top.yatt.test.javacc.ParseException: Encountered " "BAR" "BAR "" at line 1, column 5.
Was expecting:
    "DUMMY" ...
```

The output seems to be very normal. Branch1 is wrongly chosen immediately, and we should chose Branch2.

I am feeling that I have tried but I couldn't explain the problem smoothly in our discussion by using grammar and input. Could you please confirm whether the code snippet in ParseEngine is right? (I don't really want to use this way, but I failed using other ways):

```java
 for (int i = 0; min > 1 && i < e_nrw.getChoices().size(); i++) { 
   nested_e = (Expansion)(e_nrw.getChoices().get(i)); 
   int min1 = minimumSize(nested_e, min); 
   if (min > min1) min = min1; 
 } 
```

The code is to judge the MINIMAL POSSIBLE lookahead consumption for `Choice`. Please look at the loop condition `min > 1`, the condition causes the loop break up with unpredictable value of `min`. Say we have choices (I use a number to represent the minimum size of each choice):

1. ( 1 | 1 )
2. ( 2 | 2 )
3. ( 1 | 2 | 3 | 4 )
4. ( 1 | 0 )
5. ( 1 | 0 | 3 | 4)
6. ( 3 | 1 | 0 | 4)
7. ( 1 | 2 | 0 | 4 )

Can someone tell the minimum size generated for each choice? In current {{ParseEnging}}, it will generate 1, 2, 1, 1, 1, 1, 0 respectively. Are the **1**s for choice 4, 5, 6 really right? I can't imagine how and when a logic like this make any sense. When we want a minimal, we should just get the minimal. If there is a 0, we just get the 0. I don't think there is any "subtle" things behind this. Using 0 doesn't break anything, generate more code for a lookahead method doesn't break anything, it just ensure the method is safe for parsing. The previous code generated is simply not safe.

**Comment by new-javacc** [comment]
Your assertion that we should choose branch2 is not correct. I don't understand what "safe for parsing" means. Javacc semantics for lookahead clearly say this isn't expected behavior.The general rule in javacc is to use lookahead only for common prefixes.So I say this is clearly not a bug.Your work around is to put more deterministic choice bar bar first.On May 31, 2019 9:54 PM, Hongze Zhang <notifications@github.com> wrote:The debug output for following grammar and input:
grammar
void CompilationUnit() : {}
{
    ( LOOKAHEAD(2) Branch1() | Branch2() )
}

void Branch1() : {}
{
    ( <FOO> | {} ) <BAR> <DUMMY>
}

void Branch2() : {}
{
    <BAR> <BAR>
}


input
BAR BAR

debug output

Call:	CompilationUnit
  Call:	Branch1(LOOKING AHEAD...)
    Visited token: <"BAR" at line 1 column 1>; Expected token: <"FOO">
    Visited token: <"BAR" at line 1 column 1>; Expected token: <"BAR">
  Return: Branch1(LOOKAHEAD SUCCEEDED)
  Call:	Branch1
    Consumed token: <"BAR" at line 1 column 1>
  Return: Branch1
Return: CompilationUnit
Exception in thread "main" top.yatt.test.javacc.ParseException: Encountered " "BAR" "BAR "" at line 1, column 5.
Was expecting:
    "DUMMY" ...

The output seems to be very normal. Branch1 is wrongly chosen immediately, and we should chose Branch2.
I am feeling that I have tried but I couldn't explain the problem smoothly in our discussion by using grammar and input. Could you please confirm whether the code snippet in ParseEngine is right? (I don't really want to use this way, but I failed using other ways):
 for (int i = 0; min > 1 && i < e_nrw.getChoices().size(); i++) { 
   nested_e = (Expansion)(e_nrw.getChoices().get(i)); 
   int min1 = minimumSize(nested_e, min); 
   if (min > min1) min = min1; 
 } 
The code is to judge the MINIMAL POSSIBLE lookahead consumption for Choice. Please look at the loop condition min > 1, the condition causes the loop break up with unpredictable value of min. Say we have choices (I use a number to represent the minimum size of each choice):
( 1 | 1 )( 2 | 2 )( 1 | 2 | 3 | 4 )( 1 | 0 )( 1 | 0 | 3 | 4)( 3 | 1 | 0 | 4)( 1 | 2 | 0 | 4 )
Can someone tell the minimum size generated for each choice? In current {{ParseEnging}}, it will generate 1, 2, 1, 1, 1, 1, 0 respectively. Are the 1s for choice 4, 5, 6 really right? I can't imagine how and when a logic like this make any sense. When we want a minimal, we should just get the minimal. If there is a 0, we just get the 0. I don't think there is any "subtle" things behind this. Using 0 doesn't break anything, generate more code for a lookahead method doesn't break anything, it just ensure the method is safe for parsing. The previous code generated is simply not safe.

**Comment by zhztheplayer** [comment]
@new-javacc "Safe" means when we use LOOKAHEAD(2), we should make sure we checked at least 2 tokens in method jj_xxx(). Method jj_xxx() generated for grammer `LOOKAHEAD(2) ( <FOO> | {} ) <BAR> <DUMMY>`  doesn't check for the tail token \<DUMMY>, and \<DUMMY> can happen to be the second token of input token stream, we must make sure 2 tokens are checked, so the method is not safe. Where for `LOOKAHEAD(2) ( <FOO> <FOO> | {} ) <BAR> <DUMMY>`, the method checks \<DUMMY>, it's safe.

Again, I strongly recommend you can take some time to analyze the code:

```java
 for (int i = 0; min > 1 && i < e_nrw.getChoices().size(); i++) { 
   nested_e = (Expansion)(e_nrw.getChoices().get(i)); 
   int min1 = minimumSize(nested_e, min); 
   if (min > min1) min = min1; 
 } 
```

I think I had explained many times why there is a bug in such code. It is pretty straightforward, `min > 1` is a wrong loop condition. When you require for a min value you should either use `min > 0`, or just remove the loop condition. Using `min > 1` will NOT give you the actual min value. It just immediately return when it sees **1** or **0**. We can't treat a **1** to be the min value without checking all choices, because there might be some **0**s on the rhs of **1**.

**Comment by new-javacc** [comment]
I don't want to make a change that doesn't have a test case especially 
since this worked 23+ years now without problems. The case you gave is 
not a valid bug.

And from below:

"at least"

that's totally against JavaCC design. We are not changing it now. It's 
defined to be "as many as required upto 2". The right way to solve your 
problem is to put bar bar as the first production.

**Comment by zhztheplayer** [comment]
> And from below: "at least" that's totally against JavaCC design.

I meant, we should make sure we generate la code for satisfying the requirement of checking at least 2 tokens for LOOKAHEAD(2). We sometimes only need to check < 2 tokens, that is very normal. But when we have 2 tokens need to be checked (e.g. \<BAR> \<BAR>), we should ensure we have generated enough la code for that. Currently we don't.

> The right way to solve your problem is to put bar bar as the first production.

The workaround is not possible to use for very complex grammars. Could you please consider the following:

```
void Branch1() : {}
{
    ( <FOO> | {} ) <BAR> <DUMMY>
}

void Branch2() : {}
{
    ( <FOO> | {} ) <BAR> <BAR>
}

```

If we use garmmar  ` ( LOOKAHEAD(2) Branch1() | Branch2() )`, it can't parse `BAR BAR`. If we use grammar ` ( LOOKAHEAD(2) Branch2() | Branch1() )`, it can't parse `BAR DUMMY`.

**Comment by new-javacc** [comment]
Your last statement with two options is the essence of JavaCC. You need 
to left factor out your grammar because it simply takes the first choice 
period and {} does not require look ahead so it will take that option. 
The right way to your grammar is:

[ <FOO> ] <BAR> ( <BAR | <DUMMY> )

That's cleanest, most readable  way of writhing this rule.

**Comment by zhztheplayer** [comment]
Please consider: 
```
void Branch1() : {}
{
    ( <FOO1> | [ <FOO3> ] ) <BAR> <DUMMY>
}

void Branch2() : {}
{
    ( <FOO2> | ( <FOO4> )* ) <BAR> <BAR>
}
```

**Comment by new-javacc** [comment]
Same issue - it's not a test case for the "bug". You need to factor out 
left side because this a top-down parser and write a javacc grammar 
properly taking into account how javacc works. It will simply take the 
first choice it can with the minimal lookahead - so *, [] and empty 
choice {} are always taken. So you need to rework your grammar to make 
sure you don't have crazy combinations in the prefix especially 
optional/empty. And anyway these kinds of grammars hard to understand 
and maintain. So if you have real grammar with this issue, send it over 
and we can see what can be done.

**Comment by zhztheplayer** [comment]
Grammar `LOOKAHEAD(2) ( <FOO1> | ( <FOO2> )* ) <BAR> <DUMMY>` accepts `BAR BAR`, but -
grammar `LOOKAHEAD(2) ( <FOO1> <FOO1> | ( <FOO2> )* ) <BAR> <DUMMY>` rejects `BAR BAR`. That is the bug. They either both accept, or both reject.

**Comment by zhztheplayer** [comment]
> *, [] and empty choice {} are always taken. 

And I think I understood your concern above. `( <FOO> )*` / `[ <FOO> ]` / `{}` are all min=0 expansions, they can always be taken, that's no problem. But we can't consider that they consume one lookahead count, they should all consume zero. We can take `[ <FOO> ]`, but since we don't have any  `<FOO>`s, we should keep the rest lookahead count numbers and consume them afterwards. Decreasing the number by **1** is incorrect.

**Comment by new-javacc** [comment]
So the issue is you don't know that. JavaCC doesn't let you specify 
"just empty" as a production so at least one of the choices has to have 
 > 0 tokens. That's why to avoid infinite looping, we do this. Your 
example is too simplistic. If have recursion, this will loop because 
lookahead will never go down and cause major problems. It's been a long 
time (> 20yrs :)) but now I remember this.

**Comment by zhztheplayer** [comment]
> this will loop because lookahead will never go down and cause major problems.

The concern is not valid - infinite looping will never happen. JavaCC's lookahead consuming is based on how many tokens have been checked. The number of tokens in input stream are always limited, as a result a lookahead call always has an end.

Although I know the code stands there for a long time, I still know a bug is there. Fixes for bugs that last several decades in opensource projects are not that rare. So, it would be great if someone can show some bad cases against the proposed fix for this issue, such as "infinite loop" or something. I'll do analysis in detail.

**Comment by new-javacc** [comment]
Note that javacc also has semantic lookahead which is actually used during lookahead (unlike the syntactic one). And also there is the indefinite lookahead where you can actually lookahead a production. So I nee to look at all combinations. And it looks so specific for > 1 I have a feeling it's intentional and not a bug.And most importantly you havr a workaround - just use lookahead 3 and it will just work. On Jun 2, 2019 1:05 AM, Hongze Zhang <notifications@github.com> wrote:
this will loop because lookahead will never go down and cause major problems.

The concern is not valid - infinite looping will never happen. JavaCC's lookahead consuming is based on how many tokens have been checked. The number of tokens in input stream are always limited, as a result a lookahead call always has an end.
Although I know the code stands there for a long time, I still know a bug is there. Fixes for bugs that last several decades in opensource projects are not that rare. So, it would be great if someone can show some bad cases against the proposed fix for this issue, such as "infinite loop" or something. I'll do analysis in detail.
